File: compstat-rio-ai-api/app/services/risk.py
from typing import Any, Literal


RiskLevel = Literal["baixo", "médio", "alto", "crítico"]
PriorityLevel = Literal["Baixa", "Média", "Alta", "Emergencial"]


def calculate_risk_score(segment: dict[str, Any]) -> int:
    return int(
        segment["crime_score"]
        + segment["urban_factor_score"]
        + segment["intelligence_score"]
        + segment["temporal_score"]
    )


def classify_risk(total_score: int) -> RiskLevel:
    if total_score > 100:
        return "crítico"

    if total_score >= 70:
        return "alto"

    if total_score >= 40:
        return "médio"

    return "baixo"


def calculate_bingo(segment: dict[str, Any]) -> bool:
    layer_scores = [
        segment["crime_score"],
        segment["urban_factor_score"],
        segment["intelligence_score"],
        segment["temporal_score"],
    ]
    elevated_layers = [
        layer_score for layer_score in layer_scores if int(layer_score) > 20
    ]

    return len(elevated_layers) >= 3


def priority_from_risk(risk_level: RiskLevel) -> PriorityLevel:
    priority_by_risk: dict[RiskLevel, PriorityLevel] = {
        "baixo": "Baixa",
        "médio": "Média",
        "alto": "Alta",
        "crítico": "Emergencial",
    }

    return priority_by_risk[risk_level]


def enrich_segments(segments: list[dict[str, Any]]) -> list[dict[str, Any]]:
    enriched_segments: list[dict[str, Any]] = []

    for segment in segments:
        total_score = calculate_risk_score(segment)
        risk_level = classify_risk(total_score)
        enriched_segments.append(
            {
                **segment,
                "total_score": total_score,
                "totalScore": total_score,
                "risk_level": risk_level,
                "riskLevel": risk_level,
                "priority": priority_from_risk(risk_level),
                "bingo": calculate_bingo(segment),
            }
        )

    return enriched_segments


def find_priority_segment(segments: list[dict[str, Any]]) -> dict[str, Any]:
    return sorted(
        segments,
        key=lambda segment: int(segment["total_score"]),
        reverse=True,
    )[0]

File: compstat-rio-ai/lib/risk.ts
export type RiskLevel = "baixo" | "médio" | "alto" | "crítico";
export type PriorityLevel = "Baixa" | "Média" | "Alta" | "Emergencial";

export type RiskLayerScores = {
  crime_score: number;
  urban_factor_score: number;
  intelligence_score: number;
  temporal_score: number;
};

export type AreaSegment = RiskLayerScores & {
  name: string;
};

export type EnrichedAreaSegment = AreaSegment & {
  totalScore: number;
  riskLevel: RiskLevel;
  priority: PriorityLevel;
  bingo: boolean;
};

export function calculateRiskScore(segment: RiskLayerScores): number {
  return (
    segment.crime_score +
    segment.urban_factor_score +
    segment.intelligence_score +
    segment.temporal_score
  );
}

export function classifyRisk(totalScore: number): RiskLevel {
  if (totalScore > 100) {
    return "crítico";
  }

  if (totalScore >= 70) {
    return "alto";
  }

  if (totalScore >= 40) {
    return "médio";
  }

  return "baixo";
}

export function calculateBingo(segment: RiskLayerScores): boolean {
  const elevatedLayers = [
    segment.crime_score,
    segment.urban_factor_score,
    segment.intelligence_score,
    segment.temporal_score
  ].filter((layerScore) => layerScore > 20);

  return elevatedLayers.length >= 3;
}

export function priorityFromRisk(riskLevel: RiskLevel): PriorityLevel {
  const priorityByRisk: Record<RiskLevel, PriorityLevel> = {
    baixo: "Baixa",
    médio: "Média",
    alto: "Alta",
    crítico: "Emergencial"
  };

  return priorityByRisk[riskLevel];
}

export function enrichSegments(segments: AreaSegment[]): EnrichedAreaSegment[] {
  return segments.map((segment) => {
    const totalScore = calculateRiskScore(segment);
    const riskLevel = classifyRisk(totalScore);

    return {
      ...segment,
      totalScore,
      riskLevel,
      priority: priorityFromRisk(riskLevel),
      bingo: calculateBingo(segment)
    };
  });
}

export function findPrioritySegment(
  segments: EnrichedAreaSegment[]
): EnrichedAreaSegment {
  return [...segments].sort(
    (firstSegment, secondSegment) =>
      secondSegment.totalScore - firstSegment.totalScore
  )[0];
}
